Created flimesolve.py

magnamancer opened the pull request:
**Description**
Added a new open system solver based on Floquet theory

Ericgig reviewed (changes requested):
Something is strange in `floquet.py`, did you include change not meant to be here?
Both `floquet.py` and `flimesolve.py` have the original `fmesolve` and the new `flimesolve`...

Tests should pass.
Right now, scipy 1.11 breaks cvxpy which we use so this breaks a few tests, but the changes you made to correlation.py also break the tests.
We will want tests for the new solver. 

Could you also review the style (pep8).
- There are some very long line, we use the official 80 characters width.
- space are not uniform, sometime they are doubled, other time missing.
- There are rules for empty lines, there should not be 3 empty lines in a function.

magnamancer commented:
Hey Eric,

Looks like I did mess up with the floquet.py versus flimesolve.py. I think
I forgot at some point that I was to include all my new stuff in just the
flimesolve.py, and then forgot about that script altogether. I apologize
for this, and I'll fix it.

I'll also go back through and fix style issues. As with above, I think I
forgot to keep the proper styling at a certain point.

I tried to keep my changes to correlation.py to a minimum, and contained to
simply adding in the new solver in the make_solver function, as well as
adding in some inputs as options to that solver, but I'll go through and
review to see what's wrong.

Thanks for your feedback, and I'll try to get this all fixed up as soon as
I can.


On Mon, Jun 26, 2023, 4:49 PM Eric Giguère ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Something is strange in floquet.py, did you include change not meant to
> be here?
> Both floquet.py and flimesolve.py have the original fmesolve and the new
> flimesolve...
>
> Tests should pass.
> Right now, scipy 1.11 breaks cvxpy which we use so this breaks a few
> tests, but the changes you made to correlation.py also break the tests.
> We will want tests for the new solver.
>
> Could you also review the style (pep8).
>
>    - There are some very long line, we use the official 80 characters
>    width.
>    - space are not uniform, sometime they are doubled, other time missing.
>    - There are rules for empty lines, there should not be 3 empty lines
>    in a function.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qutip/qutip/pull/2140#pullrequestreview-1499406717>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AV5WL5EA3UU23BHDUONIEN3XNHYVNANCNFSM6AAAAAAWLGP4AU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

magnamancer commented:
Hey Eric,

I'm going to submit another pull request, and it should be much better this
time. I made these changes:

   - Fixed whitespace, blank line, and linting issues as far as I can tell
      - I don't have any formal code training, so I'm not sure that I broke
      lines up entirely the best way in all cases, but I think it looks okay
      enough as of now.
      - Took all of my new functions out of floquet.py and placed them
   properly into flimesolve.py
      - The version of floquet.py that I'll be submitting was one I
      directly copy-pasted from the main repository, such that there
shouldn't be
      any changes whatsoever to it
      - replaced all scipy functions with numpy equivalents
   - Fixed the issues in correlation.py, as far as I know
      - Turns out the issue here was that I never updated correlation.py to
      my most current version, which should only add in flimesolve as an option
      for solvers (with an if statement or two to get the Hamiltonian into a
      FloquetBasis object if flimesolve is the selected solver).
Apologies again
      for this.

Thanks again for your feedback!

On Mon, Jun 26, 2023 at 6:32 PM Fenton Clawson ***@***.***> wrote:

> Hey Eric,
>
> Looks like I did mess up with the floquet.py versus flimesolve.py. I think
> I forgot at some point that I was to include all my new stuff in just the
> flimesolve.py, and then forgot about that script altogether. I apologize
> for this, and I'll fix it.
>
> I'll also go back through and fix style issues. As with above, I think I
> forgot to keep the proper styling at a certain point.
>
> I tried to keep my changes to correlation.py to a minimum, and contained
> to simply adding in the new solver in the make_solver function, as well as
> adding in some inputs as options to that solver, but I'll go through and
> review to see what's wrong.
>
> Thanks for your feedback, and I'll try to get this all fixed up as soon as
> I can.
>
>
> On Mon, Jun 26, 2023, 4:49 PM Eric Giguère ***@***.***>
> wrote:
>
>> ***@***.**** requested changes on this pull request.
>>
>> Something is strange in floquet.py, did you include change not meant to
>> be here?
>> Both floquet.py and flimesolve.py have the original fmesolve and the new
>> flimesolve...
>>
>> Tests should pass.
>> Right now, scipy 1.11 breaks cvxpy which we use so this breaks a few
>> tests, but the changes you made to correlation.py also break the tests.
>> We will want tests for the new solver.
>>
>> Could you also review the style (pep8).
>>
>>    - There are some very long line, we use the official 80 characters
>>    width.
>>    - space are not uniform, sometime they are doubled, other time
>>    missing.
>>    - There are rules for empty lines, there should not be 3 empty lines
>>    in a function.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/qutip/qutip/pull/2140#pullrequestreview-1499406717>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AV5WL5EA3UU23BHDUONIEN3XNHYVNANCNFSM6AAAAAAWLGP4AU>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

-- 
Best Regards,
Fenton Clawson

Ericgig reviewed (commented):
I took a first look.
I see a lot of small style issues, I would suggest that you try a formatting tool: pycodestyle, flake8, black, etc.
I will go into detail on flimesove.py later, but would you mind if I did a pull request agins your local branch to make the tensor using qutip's types. We want to force the use of numpy to the minimum so it can work with other backend, (jax, cupy, tensorflow, etc.).

Ericgig commented:
I changed the merge target from `qutip-5.0.X` to `master`. The `qutip-...` branches are for released version.

magnamancer commented on doc/guide/dynamics/dynamics-floquet.rst:
```diff
@@ -283,3 +283,29 @@ Finally, :func:`qutip.solver.floquet.fmmesolve`  always expects the ``e_ops`` to
+-------------------------------------------
+The QuTiP function :func:`qutip.floquet.flimesolve` implements the Floquet-Lindblad master equation. It calculates the dynamics of a system given its initial state, a time-dependent Hamiltonian, a list of operators through which the system couples to its environment and a list of corresponding time-independent decay rates for the associated operators. Flimesolve is much like the :func:`qutip.meso ...
+
+.. note::
+
+    Currently the :func:`qutip.floquet.flimesolve` can accept multiple collapse operators and their associated rates
```
This was an artifact from copying over the description of fmmesolve to use as a template. Fixed

magnamancer commented on doc/guide/scripts/floquet_ex4.py:
```diff
@@ -0,0 +1,73 @@
+    
+   
+# solve the floquet-markov master equation
+output = qutip.flimesolve(
+    H, psi0, tlist, 
+    c_ops_and_rates = [[qutip.sigmax(),gamma1]], 
+    T=T,
+    args=args, options={"store_floquet_states": True}
+)
```
Somehow missed that. Deleting the extra solve

magnamancer commented:
Used the auto-format feature for PEP8 in my IDE (Spyder), then did a second check with pycodestyle and fixed all the issues it raised. Fixed the other issues you raised as well.

magnamancer commented:
Merged

> I changed the merge target from `qutip-5.0.X` to `master`. The `qutip-...` branches are for released version. Could you merge the current master branch, we added fixes for the new numpy, scipy version and it should help tests to passes.

Merged my qutip-5.0.X to master

magnamancer commented:
> I took a first look. I see a lot of small style issues, I would suggest that you try a formatting tool: pycodestyle, flake8, black, etc. I will go into detail on flimesove.py later, but would you mind if I did a pull request agins your local branch to make the tensor using qutip's types. We want to force the use of numpy to the minimum so it can work with other backend, (jax, cupy, tensorflow, etc.).

Please feel free to do a pull request. Alternatively, I can spend some time trying to figure it out myself. Not trying to add anything onto your workload.
